module/Company/src/Company/Controller/AdminController.php

```diff
@@ -397,6 +471,54 @@ public function deleteCompanyAction()
         return $this->notFoundAction();
     }
 
+    public function addCategoryAction()
+    {
+        // Get useful stuff
+        $companyService = $this->getCompanyService();
+        $categoryForm = $companyService->getCategoryForm();
+
+        // Handle incoming form results
+        $request = $this->getRequest();
+        if ($request->isPost()) {
+            // Check if data is valid, and insert when it is
+            $categories = $companyService->insertCategoryByData(
+                $request->getPost(),
+                $request->getFiles()
+            );
+            if (!is_null($categories)) {
+                // Redirect to edit page
+                return $this->redirect()->toRoute(
+                    'admin_company/default',
+                    [
+                        'action' => 'editCategory',
+                        'slugCompanyName' => $categories['nl']->getLanguageNeutralId(),
+                    ]
+                );
+            }
+        }
+
+        // The form was not valid, or we did not get data back
+
+        // Initialize the form
+        $categoryForm->setAttribute(
+            'action',
+            $this->url()->fromRoute(
+                'admin_company/default',
+                ['action' => 'addCategory']
+            )
+        );
+        $languages = $companyService->getLanguages();
+        $f = function ($lang) use ($companyService) {
+            return [$lang, $companyService->getLanguageDescription($lang),];
+        };
+        $languageDescriptions = array_map($f, $languages);
```

Comment:
Wouldn't it be better to just add a function in the service which returns this array? These 5 lines seem to be repeating a lot.